tutorials/editor/upgrading_to_godot_4.rst

```diff
@@ -0,0 +1,540 @@
+    `Software OpenGL implementations <https://github.com/pal1000/mesa-dist-win>`__
+    can be used to bypass this limitation, but they're too slow for gaming.
+
+Running the project upgrade tool
+--------------------------------
+
+.. warning::
+
+    **Make a full backup of your project** before upgrading! The project upgrade
+    tool will *not* perform any backups of the project that is being upgraded.
+
+    You can backup a project by using version control, or by copying the project
+    folder to another location.
+
+To use the project upgrade tool:
+
+1. Open the Godot 4 project manager.
+2. Import the Godot 3.x project using the **Import** button, or use the **Scan**
+   button to find the project within a folder.
+3. Double-click the imported project (or select the project then choose **Edit**).
+4. You will see a dialog appearing with two options: **Convert project.godot
+   Only** and **Convert Full Project**. After ensuring your project is backed up
+   (see the above warning), choose **Convert Full Project**. **Convert
+   project.godot Only** is intended to be used for advanced use cases *only*, in
+   case the conversion tool fails.
+5. Wait until the project conversion process finishes. This can take up to a few
+   minutes for large projects with lots of scenes.
+6. When the project manager interface becomes available again, double-click the
+   project (or select the project then choose **Edit**) to open it in the
+   editor.
+
+.. note::
+
+    Only Godot 3.0 and later projects can be upgraded using the project
+    conversion tool found in the Godot 4 editor.
+
+    Projects made using Godot versions older than 3.0 must follow a 2-step
+    process:
+
+    1. Upgrade the project to Godot 3 using the **Tools > Export to Godot 3.0**
+       option in the Godot 2.1.6 editor. Open the upgraded project in the Godot
+       3.x editor, then manually fix the project to make sure it looks and
+       runs correctly.
+    2. Only *after* performing step 1, you may upgrade the project to Godot 4.
+
+Fixing the project after running the project upgrade tool
+---------------------------------------------------------
+
+After upgrading the project, you may notice that certain things don't look as
+they should. Scripts will likely contain various errors as well (possibly
+hundreds in large projects). This is because the project upgrade tool cannot
+cater to all situations. Therefore, a large part of the upgrade process remains
+manual.
+
+Automatically renamed nodes and resources
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The list below refers to the nodes' new names in Godot 4.0. In the transition
+between Godot 3.x and 4.0, dozens of nodes were renamed. 3D nodes had a ``3D``
+suffix added to them for consistency with their 2D counterparts. This conversion
+is automatic.
+
+For ease of searching, this table lists all nodes and resources that were
+renamed in a way that does not involve only adding a ``3D`` suffix to the old name:
```

Comment:
```suggestion
The list below refers to nodes which were simply renamed for consistency or
clarity in Godot 4.0. The project upgrade tool renames them automatically in
your scripts.

One noteworthy set of renames is 3D nodes, which all got a ``3D`` suffix added for
consistency with their 2D counterparts. For example, ``Area`` is now ``Area3D``.

For ease of searching, this table lists all nodes and resources that were renamed
and are automatically converted, excluding the ones which only involved adding
a ``3D`` suffix to the old name:
```

Reply:
I tried to make the language clearer that this is *not* all renamed nodes. IMO it's still a bit confusing though to have an "exhaustive" list that actually is only a subset.

There would need to be an intro paragraph to explain how the document is structured further (trivial renames handled by the converter, API changes which require manual handling, etc.).

But maybe there should be a table which actually lists *everything*, and has a third column that clarifies where the rename is handled by the conversion tool or not (and if not, why - different API, risk of false positives, etc.).
So users could always find an old or new name with Ctrl+F.

Reply:
The table would be very long and could get outdated quickly – I guess we could find a way to generate it automatically and put it on a separate page.

Reply:
> But maybe there should be a table which actually lists everything, and has a third column that clarifies where the rename is handled by the conversion tool or not (and if not, why - different API, risk of false positives, etc.).
So users could always find an old or new name with Ctrl+F.

Absolutely. 100%. It's part of the reason as to why I expressed this:
> I was considering reordering that whole list 'cause it very much is... such a mess to read,